### PR TITLE
[ASCII-2221] GetCatalog is fx entry point, remove fx from Collectors

### PR DESCRIPTION
**Pull request by dustmop**
### What does this PR do?

This PR sets up follow-up work that will turn the Catalog into a full fledged Component. Instead of having multiple ways to select catalogs (build tags / agent flavor / etc) we will use separate implementations, which each agent binary will be able to select at its entry point. This will enable full separation of dependencies for different collector catalogs.

This PR is best reviewed one commit at a time.

### Motivation

RFC on component's new usage of fx, including how to define components with multiple implementations, as a way to control dependencies: https://docs.google.com/document/d/1sPXUviG-RvMCUvqdwfUYbTOTV05Z_y9MuwlYcyOdG0I/edit#heading=h.o5gstqo08gu5

### Additional Notes

As a next step, the catalogs will move from `comp/core/workloadmeta/collectors/catalog-{*}` to `comp/core/wmcatalog/impl-{*}` as wmcatalog becomes a true component. As for the collectors, it seems like the best option would be to move them from `comp/core/workloadmeta/collectors/internal` to `comp/core/wmcatalog/internal/collectors` in order to work with both the multi-impl paradigm and the semantics of importing internal packages.

### Possible Drawbacks / Trade-offs

It may be more difficult to pass arbitrary data to the Collector's constructor, as the remoteworkloadmeta collector shows. However, it does not seem to be the case that Collectors need to do this very often; most simply require the config.

### Describe how to test/QA your changes

No functional change, behavior should remain the same as before this refactoring.

### CODE REVIEW

**Comment by GustavoCaso** [comment]
I have suggestion to make to refactor a bit how collectors initialization works.

I see we do not use the error returned from the constructors. At the same time the `Start` function from the Collector usually check if the current environments support such collector: 

```go
if !pkgconfig.IsFeaturePresent(pkgconfig.Docker) {
  return errorspkg.NewDisabled(componentName, "Agent is not running on Docker")
}
```

I say we could move the checking of the feature to the `NewCollector` collector function, then we are building the collectors catalog, we check the error returned from the `newCollector` function and if is `nil` we add the catalog. 

```go
// NewCollector returns a new docker collector
func NewCollector(cfg config.Component) (wmcatalog.Collector, error) {
	if !pkgconfig.IsFeaturePresent(pkgconfig.Docker) {
		return nil, errorspkg.NewDisabled(componentName, "Agent is not running on Docker")
	}
	
	return &collector{
		id:      collectorID,
		config:  cfg,
		catalog: workloadmeta.NodeAgent | workloadmeta.ProcessAgent,
	}, nil
}
```

Then in the util build catalog would be:

```go
// BuildCatalog builds a list of Collectors by invoking their constructors
func BuildCatalog(cfg config.Component, constructors ...func(config.Component) (wmcatalog.Collector, error)) []wmcatalog.Collector {
	results := []wmcatalog.Collector{}
	for _, ctor := range constructors {
		item, err := ctor(cfg)
		if err != nil {
			results = append(results, item)
		}
	}
	return results
}
``` 

Positive outcomes, is that we instantiate less objects 🎉

File: comp/core/workloadmeta/collectors/internal/ecs/ecs.go [hunk 1]
```diff
@@ -68,33 +62,26 @@ type resourceTags struct {
...
 }
 
-// GetFxOptions returns the FX framework options for the collector
-func GetFxOptions() fx.Option {
-	return fx.Provide(NewCollector)
-}
-
 func (c *collector) Start(ctx context.Context, store workloadmeta.Component) error {
-	if !pkgConfig.IsFeaturePresent(pkgConfig.ECSEC2) {
+	if !pkgconfig.IsFeaturePresent(pkgconfig.ECSEC2) {
```
Comment (GustavoCaso) [inline comp/core/workloadmeta/collectors/internal/ecs/ecs.go:84]: Nitpick: We could remove the usage of `pkgconfig` and use `pkg/config/env` to get the same outcome:

`!env.IsFeaturePresent(env.ECSEC2)`

**Comment by GustavoCaso** [comment]
Are we planning on removing the `remotewmonly` build flag on the next steps?

File: comp/core/wmcatalog/def/collector.go [hunk 2]
```diff
@@ -3,20 +3,21 @@
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
 // Copyright 2016-present Datadog, Inc.
 
-package workloadmeta
+// Package wmcatalog defines the catalog of collectors for workloadmeta
+package wmcatalog
```
Comment (GustavoCaso) [inline comp/core/wmcatalog/def/collector.go:8]: Do you plan on adding a `Component` interface to this component?

**Review by vickenty — APPROVED** [review]

File: comp/core/workloadmeta/collectors/internal/ecs/ecs.go (hunk 1 above)
Comment (dustmop) [reply comp/core/workloadmeta/collectors/internal/ecs/ecs.go:84]: Ah, I didn't realize that they were aliased from the env package.

File: comp/core/wmcatalog/def/collector.go (hunk 2 above)
Comment (dustmop) [reply comp/core/wmcatalog/def/collector.go:8]: Yes, as mentioned in the PR description.

**Comment by dustmop** [comment]
> Are we planning on removing the `remotewmonly` build flag on the next steps?

Yes, the wmcatalog component will have its own implementation that takes over what the `remotewmonly` build flag is doing. I checked with @paulcacheux a while back and heard that doing so is okay.

**Comment by dustmop** [comment]
> I see we do not use the error returned from the constructors. At the same time the `Start` function from the Collector usually check if the current environments support such collector:

Ah, that's a good point that the error returned by the constructors isn't being used. It had semantics that fx cares about, but this is removing fx from the collectors, which is introducing a bug. I'll work on fixing that.

**Comment by dustmop** [comment]
> > I see we do not use the error returned from the constructors. At the same time the `Start` function from the Collector usually check if the current environments support such collector:
> 
> Ah, that's a good point that the error returned by the constructors isn't being used. It had semantics that fx cares about, but this is removing fx from the collectors, which is introducing a bug. I'll work on fixing that.

Looking into it, the error should be removed from the collector constructors, it can't be used and we're already checking for whether the collector is non-nil.

File: comp/core/workloadmeta/collectors/catalog-less/options.go [hunk 3]
```diff
@@ -24,27 +24,35 @@ import (
...
-		firstArg(remoteworkloadmeta.NewCollector(cfg)),
-		// TODO: remoteworkloadmetaParams(),
-		firstArg(processcollector.NewCollector(cfg)),
-		firstArg(host.NewCollector(cfg)),
+	var filter *workloadmeta.Filter // Nil filter accepts everything
+
+	// Security Agent is only interested in containers
+	// TODO: (components) create a Catalog component, the implementation used by
+	// security-agent can use this filter, instead of needing to check agent.flavor
+	if flavor.GetFlavor() == flavor.SecurityAgent {
```
Comment (ogaca-dd) [inline comp/core/workloadmeta/collectors/catalog-less/options.go:38]: Alternative idea: `getCollectorList` could take a `Params` argument  which has a field `extraCollectorKind workloadmeta.Kind` that adds non default collectors. Using `Params` is the standard way to configure component.

**Review by ogaca-dd — COMMENTED** [review]
Great PR. Thanks!

**Review by brycekahle — APPROVED** [review]

**Review by robertjli — APPROVED** [review]

**Pull request closed without merging**